From b77686db1728a27525f82fd32e2506171d5491f8 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 25 Aug 2014 07:53:03 -0400
Subject: [tools] perf/symbols: Load map before using map->map_ip()

Message-id: <1408953210-26343-296-git-send-email-jolsa@redhat.com>
Patchwork-id: 89016
O-Subject: [PATCH RHEL7.1 BZ1133083 295/322] perf symbols: Load map before using map->map_ip()
Bugzilla: 1133083
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1133083
https://bugzilla.redhat.com/show_bug.cgi?id=1133083

upstream
========
commit 4afc81cd1caa93daa50c1c29a3ab747c978abc13
Author: Masami Hiramatsu <masami.hiramatsu.pt@hitachi.com>
Date: Thu Jan 23 02:29:50 2014 +0000

description
===========
In map_groups__find_symbol() map->map_ip is used without ensuring the
map is loaded. Then the address passed to map->map_ip isn't mapped at
the first time.

E.g. below code always fails to get a symbol at the first call;

 addr = /* Somewhere in the kernel text */
 symbol_conf.try_vmlinux_path = true;
 symbol__init();
 host_machine = machine__new_host();
 sym = machine__find_kernel_function(host_machine,
      addr, NULL, NULL);
 /* Note that machine__find_kernel_function calls
    map_groups__find_symbol */

This ensures it by calling map__load before using it in
map_groups__find_symbol().
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/util/map.c b/tools/perf/util/map.c
index ee1dd68..3b97513 100644
--- a/tools/perf/util/map.c
+++ b/tools/perf/util/map.c
@@ -386,7 +386,8 @@ struct symbol *map_groups__find_symbol(struct map_groups *mg,
 {
  struct map *map = map_groups__find(mg, type, addr);
 
- if (map != NULL) {
+ /* Ensure map is loaded before using map->map_ip */
+ if (map != NULL && map__load(map, filter) >= 0) {
   if (mapp != NULL)
    *mapp = map;
   return map__find_symbol(map, map->map_ip(map, addr), filter);
-- 
1.7.1